### ISSUE TITLE
Fastapi

### ISSUE THREAD

**Issue by captain-sal** [captain-sal/cybersecurity#2]
MODEL.PY

import torch
import torch.nn as nn

class UserEmbeddingModel(nn.Module):
    def __init__(self, input_dim, embedding_dim, num_users, user_embedding_dim):
        super(UserEmbeddingModel, self).__init__()
        self.user_embedding = nn.Embedding(num_users, user_embedding_dim)
        self.lstm = nn.LSTM(input_dim + user_embedding_dim, embedding_dim, batch_first=True)
        self.fc = nn.Linear(embedding_dim, 10)  # Example output dimension

    def forward(self, x, user_id):
        user_embed = self.user_embedding(user_id).unsqueeze(1)
        user_embed = user_embed.expand(-1, x.size(1), -1)
        x = torch.cat((x, user_embed), dim=2)
        x, _ = self.lstm(x)
        x = self.fc(x[:, -1, :])
        return x

def get_model(input_dim, embedding_dim, num_users, user_embedding_dim):
    return UserEmbeddingModel(input_dim, embedding_dim, num_users, user_embedding_dim)

TRAIN_USER_MODEL.PY

import torch
from torch.utils.data import DataLoader, Dataset
import pandas as pd
from model import get_model
import os

class UserDataset(Dataset):
    def __init__(self, data_path):
        self.data = pd.read_csv(data_path)
        self.users = self.data['user_id'].unique()
        self.user_map = {user: idx for idx, user in enumerate(self.users)}

    def __len__(self):
        return len(self.data)

    def __getitem__(self, idx):
        row = self.data.iloc[idx]
        user_id = self.user_map[row['user_id']]
        x = torch.tensor(row.drop(['user_id', 'label']).values, dtype=torch.float32)
        y = torch.tensor(row['label'], dtype=torch.long)
        return x, user_id, y

def train_model(data_path, num_epochs=10, batch_size=32, learning_rate=0.001):
    dataset = UserDataset(data_path)
    dataloader = DataLoader(dataset, batch_size=batch_size, shuffle=True)

    input_dim = len(dataset.data.columns) - 2
    embedding_dim = 64
    user_embedding_dim = 32
    num_users = len(dataset.users)
    model = get_model(input_dim, embedding_dim, num_users, user_embedding_dim)

    criterion = torch.nn.CrossEntropyLoss()
    optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)

    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    model.to(device)

    for epoch in range(num_epochs):
        model.train()
        total_loss = 0
        for batch in dataloader:
            x, user_id, y = batch
            x, user_id, y = x.to(device), user_id.to(device), y.to(device)

            optimizer.zero_grad()
            output = model(x.unsqueeze(1), user_id)
            loss = criterion(output, y)
            loss.backward()
            optimizer.step()

            total_loss += loss.item()

        print(f'Epoch {epoch + 1}, Loss: {total_loss / len(dataloader):.4f}')

    os.makedirs('outputs/models', exist_ok=True)
    torch.save(model.state_dict(), 'outputs/models/shared_model.pth')

if __name__ == '__main__':
    train_model('data/processed/dataset.csv')

LOAD_USER_MODEL.PY

import torch
from model import get_model

def load_model(model_path, input_dim, embedding_dim, num_users, user_embedding_dim):
    model = get_model(input_dim, embedding_dim, num_users, user_embedding_dim)
    model.load_state_dict(torch.load(model_path))
    model.eval()
    return model

def predict(model, user_id, x):
    user_id_tensor = torch.tensor([user_id], dtype=torch.long)
    x_tensor = torch.tensor(x, dtype=torch.float32).unsqueeze(0).unsqueeze(1)
    with torch.no_grad():
        output = model(x_tensor, user_id_tensor)
    return output.argmax(dim=1).item()

Server.py

from fastapi import FastAPI, BackgroundTasks, HTTPException, Query
from typing import List
import torch
import pandas as pd
from train_user_model import train_model
from load_user_model import load_model, predict
import os

app = FastAPI()

# Load user mappings and data dimensions
data_path = 'data/processed/dataset.csv'
data = pd.read_csv(data_path)
users = data['user_id'].unique()
user_map = {user: idx for idx, user in enumerate(users)}

input_dim = len(data.columns) - 2
embedding_dim = 64
user_embedding_dim = 32
num_users = len(users)

# Load the shared model
model_path = 'outputs/models/shared_model.pth'
model = load_model(model_path, input_dim, embedding_dim, num_users, user_embedding_dim)

@app.post("/train/{user_id}/{model_name}")
def train_model_endpoint(user_id: str, model_name: str, background_tasks: BackgroundTasks):
    background_tasks.add_task(train_model, data_path, num_epochs=10, batch_size=32, learning_rate=0.001)
    return {"message": f"Training started for user {user_id} with model name {model_name}"}

@app.get("/fetch_model/{user_id}")
def fetch_model(user_id: str, feature_vector: List[float] = Query(...)):
    user_index = user_map.get(user_id)
    if user_index is None:
        raise HTTPException(status_code=404, detail=f"User {user_id} not found")

    prediction = predict(model, user_index, feature_vector)
    return {"user_id": user_id, "prediction": prediction}

@app.get("/")
def read_root():
    return {"message": "Welcome to the Model Training Server"}

if __name__ == '__main__':
    import uvicorn
    uvicorn.run(app, host='0.0.0.0', port=8000)
